design-proposals/klip-35-dynamic-processing-log-levels.md

```diff
@@ -0,0 +1,210 @@
+
+There have also been incidents where even the current error-only level of logging is too verbose. In such cases, it should be possible to turn the processing log off at runtime.
+
+In this KLIP, I propose a mechanism to dynamically control the log level of the record-processing log. Furthermore, the proposed mechanism makes it possible to change the level with a fine granularity - so you can change the level at runtime for individual queries, or even individual processing nodes of a query.
+
+For important background on the processing log, please read https://docs.ksqldb.io/en/latest/developer-guide/test-and-debug/processing-log/
+
+## In-Scope
+
+- API for getting/setting the global log level
+- API for getting/setting query/node-level log levels
+- Kafka-based stateful implementation that distributes level settings to the cluster and stores them so they are recovered on a restart.
+
+## Out-Of-Scope
+
+- Thoroughly populating the query engine with log messages to address the problems we’ve already seen. This is super important, but out of scope for this particular KLIP.
+
+## Value/Return
+
+A step toward improving UX when users observe unexpected behavior from running queries.
+
+## Public APIs
+
+### REST API
+
+### PUT /processing_logger/config
+
+Sets the root processing log configuration (currently just supports setting the log level).  This will set the level for all loggers that do not already have a level set.
+
+Request Schema:
+```
+{
+    “properties”: {
+        “level”: {“type”: “string”, “description”: “<OFF|FATAL|ERROR|WARN|INFO|DEBUG>”}
+    }
+}
+````
+
+Response Schema:
+```
+{
+    “properties”: {
+        “level”: {“type”: “string”, “description”: “The level from the request”}
+    }
+}
+```
+
+### GET /processing_logger/config
+
+Gets the current root processing log level.
+
+Response Schema:
+```
+{
+    “properties”: {
+        “level”: {“type”: “string”, “description”: “The level of the root logger”}
+    }
+}
+```
+
+### DELETE /processing_logger/config
+
+Deletes customized config for the root logger (reverts to default config as specified in the log4j config file).
+
+### PUT /processing_logger/[processing logger name]/config
+
+Sets the processing log configuration for the logger with the given processing logger name (currently just supports setting the log level). This will set the level for all child loggers (loggers that share a common name prefix) that do not already have a level set using a longer prefix.
+
+Request Fields:
+```
+{
+    “properties”: {
+        “level”: {“type”: “string”, “description”: “<OFF|FATAL|ERROR|WARN|INFO|DEBUG>”}
+    }
+}
+```
+
+Response Fields:
+```
+{
+    “properties”: {
+        “level”: {“type”: “string”, “description”: “The level from the request”}
+    }
+}
+```
+
+### GET /processing_logger/[processing logger name]/level
```

Comment:
The endpoints above are `/config` but this one and the ones below are `/level`. Were they all meant to be `/config`?

Reply:
yes! I started with it as `/level` and changed it so that it's more generic/extensible